[Bugfix] 162 layout

JaeHye0k opened the pull request:
## #️⃣연관된 이슈

> ex) #162 

## 📝작업 내용

1. 로그아웃 상태에서 커넥션 허브 페이지 접속 시 login 페이지로 리다이렉트 되는 문제 해결
2. 사이드 바 프로필 버튼 border 타원형에서 원형으로 수정
3. 사이드 바 프로필 버튼 클릭 시 로그아웃 상태일 때도 '마이페이지', '로그아웃' 표시되는 문제 해결 
4. 채팅 웹소켓 URL 로컬에서 배포 서버로 수정.
5. SRP(단일 책임 원칙)을 기반으로 SideMenu 컴포넌트 리팩토링

## 작업 상세 내용(선택)
### 1번 작업:
- 원인: 커넥션 허브 페이지 접속 시 북마크 조회 api 가 호출되고 있었음. 그러나 로그아웃 상태에서 북마크 조회를 수행할 시 401 에러가 발생하고 axios 응답 인터셉터에 의해 login 페이지로 리다이렉트됨
- 해결방법: api 호출 useQuery에 enabled:isLoggedIn 을 추가하여 로그인 되어있을 때만 api 를 호출하도록 수정함
### 4번 작업: 
- URL은 수정했지만, 실제로 배포 사이트에서 채팅 기능을 사용하려면 수정된 백엔드 코드도 백엔드 배포 서버에 반영이 되어야 합니다.
### 5번 작업
<details>
<summary>리팩토링 전, 후 비교</summary>
<div markdown="1">

### before
```tsx
// SideMenu.tsx
const SideMenu = () => {
  const navigate = useNavigate();
  const { userInfo, logout } = useAuthStore((state) => state);
  const [showLogin, setShowLogin] = useState(false);
  const {
    isOpen: isSearchModalOpen,
    openModal: openSearchModal,
    closeModal: closeSearchModal,
  } = useModal();

  const {
    messages,
    newNotification,
    markNotificationAsRead,
    setNewNotification,
  } = useNotification();
  const [showNotificationBox, setShowNotificationBox] = useState(false);
  const notificationRef = useRef<HTMLDivElement | null>(null);

  const handleNotificationClick = () => {
    setShowNotificationBox((prev) => !prev);
    setNewNotification(false);
  };

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (
        notificationRef.current &&
        !notificationRef.current.contains(event.target as Node)
      ) {
        setShowNotificationBox(false);
      }
    };

    if (showNotificationBox) {
      document.addEventListener('mousedown', handleClickOutside);
    }

    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [showNotificationBox]);

  const menuItems: {
    type: 'bell' | 'mail' | 'home' | 'search' | 'star';
    label?: string;
    onClick?: () => void;
    hasNotification?: boolean;
  }[] = [
    {
      type: 'bell',
      label: '알림',
      onClick: handleNotificationClick,
      hasNotification: newNotification,
    },
    { type: 'mail', label: '메세지', onClick: () => navigate('/chat') },
    { type: 'home', label: '피드', onClick: () => navigate('/') },
    { type: 'search', label: '검색', onClick: openSearchModal },
    {
      type: 'star',
      label: '커넥션 허브',
      onClick: () => navigate('/projects'),
    },
  ];

  return (
    <>
      {isSearchModalOpen && <SearchModal onClose={closeSearchModal} />}

      <div className='flex flex-col lg:py-[20px] justify-between items-center h-full'>
        <div className='mb-8 cursor-pointer' onClick={() => navigate('/')}>
          <Logo />
        </div>
        <Menu items={menuItems} />

        {showNotificationBox &&
          createPortal(
            <div className='fixed inset-0 flex items-center justify-center z-[1000]'>
              <div
                ref={notificationRef}
                className='absolute left-[90px] top-[50px] w-[370px] h-[700px] bg-white bg-opacity-95 rounded-xl drop-shadow-lg px-[20px] py-[20px] overflow-y-auto z-50'
              >
                <div className='flex w-full flex-col items-center gap-[10px]'>
                  <div className='text-[18px] font-semibold text-[#48484a]'>
                    알림 📫
                  </div>
                  {messages.length === 0 ? (
                    <div className='text-[16px] text-[#828282]'>
                      현재 새로운 알림이 없습니다.
                    </div>
                  ) : (
                    <div className='flex w-full flex-col gap-[20px]'>
                      {messages.map((message) => (
                        <div
                          key={message.notificationId}
                          className='flex w-full justify-start text-[14px] items-center gap-[10px]'
                        >
                          <Avatar
                            src={message.senderProfileUrl || undefined}
                            size='xs'
                          />
                          <div>
                            <div>{message.message}</div>
                            <div className='text-[12px] text-gray-500'>
                              {message.timestamp}
                            </div>
                          </div>
                          <div
                            onClick={() =>
                              markNotificationAsRead(message.notificationId)
                            }
                          >
                            <Icon
                              type='trash'
                              color='black'
                              className='w-[20px] h-[20px] cursor-pointer'
                            />
                          </div>
                        </div>
                      ))}
                    </div>
                  )}
                </div>
              </div>
            </div>,
            document.body
          )}

        <div className='relative'>
          <Avatar
            size='sm'
            alt='User Avatar'
            className='cursor-pointer border-4 border-transparent hover:border-[#c7c7c7] transition-shadow duration-300'
            src={userInfo?.profileUrl || undefined}
            onClick={() => setShowLogin((prev) => !prev)}
          />
          {showLogin &&
            (isLoggedIn ? (
              <Popup
                position='right'
                popupHandler={[
                  {
                    onClick: () => navigate(`/@${userInfo?.nickname}`),
                    text: '마이페이지',
                    icon: <Icon type='user' className='w-6' />,
                  },
                  {
                    onClick: () => logout(),
                    text: '로그아웃',
                    icon: <Icon type='logout' className='w-6' />,
                  },
                ]}
                innerClassname='top-[-30%]'
              />
            ) : (
              <Popup
                position='right'
                popupHandler={[
                  {
                    onClick: () => {
                      navigate('/login');
                      setShowLogin(false);
                    },
                    text: '로그인/회원가입',
                    icon: <Icon type='user' className='w-6' />,
                  },
                ]}
                innerClassname='top-[10px]'
              />
            ))}
        </div>
      </div>
    </>
  );
};
```

### after
```tsx
// SideMenu.tsx
const SideMenu = () => {
  return (
    <div className='flex flex-col lg:py-[20px] justify-between items-center h-full'>
      <SideMenuLogo />
      <SideMenuNavigation>
        <SideMenuNav.Bell />
        <SideMenuNav.Mail />
        <SideMenuNav.Home />
        <SideMenuNav.Search />
        <SideMenuNav.ConnectionHub />
      </SideMenuNavigation>
      <SideMenuNotification />
      <SideMenuProfile />
    </div>
  );
};
```

</div>
</details>

JaeHye0k committed:
[Fix] 로그아웃 상태에서 표시되는 팝업 메뉴 수정

기존: 로그아웃 상태에서도 마이페이지, 로그아웃 메뉴가 표시되는 문제가 있었음
수정: 로그아웃 상태에서는 로그인/회원가입 메뉴만 표시되도록 수정

JaeHye0k committed:
[Refactor] 책임에 따라 SideMenu 컴포넌트 분리

- SideMenuLogo: 로고 클릭 관련
- SideMenuNavigation: 사이드바 네비게이션 메뉴 관련
- SideMenuNotification: 알림창 관련
- SideMenuProfile: 프로필 버튼 클릭 관련

JaeHye0k committed:
[Refactor] Context API 를 이용해 관심사 분리

알림에 관한 상태를 NotificationProvider 안으로 이동시켜 props 대신 Context API를 이용하도록 수정
알림에 관한 상태를 구독하고 있던 컴포넌트들은 props 에서 Context API로만 바뀐것이기 때문에 Context API로 인한 불필요한 리렌더링은 발생하지 않음

JaeHye0k committed:
[Refactor] SRP(단일 책임 원칙)를 바탕으로 리팩토링

- SideMenuNavigation: 알림, 메시지, 홈, 검색, 커넥션 허브 메뉴 관심사 분리 및  합성 컴포넌트화
- SideMenuNotification:
    - 알림 메시지 목록 관련 로직을 NotificationList 컴포넌트로 분리
    - 외부 영역 클릭시 창 닫힘 로직을 useOutsideClick 훅으로 추상화

JaeHye0k committed:
[Fix] 로그아웃 상태에서 커넥션 허브 페이지 접속 시 login 페이지로 리다이렉트 되는 문제 해결

- 원인: 커넥션 허브 페이지 접속 시 북마크 조회 api 가 호출되고 있었음. 그러나 로그아웃 상태에서 북마크 조회를 수행할 시 401 에러가 발생하고 axios 응답 인터셉터에 의해 login 페이지로 리다이렉트됨
- 해결방법: api 호출 useQuery에 enabled:isLoggedIn 을 추가하여 로그인 되어있을 때만 api 를 호출하도록 수정함

JaeHye0k committed:
[Fix] 소켓 URL 주소 배포 서버로 수정

JaeHye0k committed:
[Merge] branch 'develop' into fix/162-layout

chaeyun-sim commented on src/components/molecules/SideMenuLogo.tsx:
```diff
@@ -0,0 +1,13 @@
+import Logo from '@/components/atoms/Logo';
+import { useNavigate } from 'react-router-dom';
+
+const SideMenuLogo = () => {
+  const navigate = useNavigate();
+  return (
+    <div className='mb-8 cursor-pointer' onClick={() => navigate('/')}>
+      <Logo />
+    </div>
+  );
+};
+
+export default SideMenuLogo;
```
이 부분 Logo 컴포넌트에 수정해도 될거 같습니다. (atoms/Logo)

JaeHye0k replied:
위 코드를 `<Logo>` 컴포넌트에 적용하고 `<SideMenuLogo>` 대신 `<Logo>` 를 직접 사용해도 될 것 같다는 내용으로 이해했는데 제가 이해한 게 맞나요?

chaeyun-sim replied:
@JaeHye0k 네네 맞습니다!

JaeHye0k committed:
[Move] optimizeImage 파일 이동 및 파일명 수정

src/constants/OptimizeImage.ts ->  src/utils/optimizeImage.ts

1. 상수가 아니라 함수이므로 constants 에서 utils 로 이동
2. 컴포넌트가 아니므로 파일명 첫 글자를 소문자로 작성

JaeHye0k committed:
[Merge] branch 'develop' into fix/162-layout
